servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/SpliceFlatStreamToMetaSingle.java

```diff
@@ -196,9 +198,15 @@ public void onNext(@Nullable Object obj) {
                 metaSeenInOnNext = true;
                 final Data data;
                 try {
-                    data = parent.packer.apply(meta, maybePayloadSubUpdater.compareAndSet(this, null, PENDING) ?
-                            newPayloadPublisher() : Publisher.failed(StacklessCancellationException.newInstance(
-                                    "Canceled prematurely from Data", SplicingSubscriber.class, "cancelData(..)")));
+                    final Publisher<Payload> payload;
+                    if (maybePayloadSubUpdater.compareAndSet(this, null, PENDING)) {
+                        payload = newPayloadPublisher();
+                    } else {
+                        assert maybePayloadSub == CANCELED;
+                        maybePayloadSubUpdater.compareAndSet(this, CANCELED, EMPTY_COMPLETED_DELIVERED);
+                        payload = Publisher.failed(newCancellationException());
```

Comment:
I believe this `cas` should never fail: can we throw an AssertionError or similar if it does?

Reply:
good idea, it should never fail